File: src/spirite/base/graphics/gl/GLImage.java
package spirite.base.graphics.gl;

import com.jogamp.opengl.GL2;
import spirite.base.graphics.RawImage;
import spirite.base.util.glu.GLC;
import spirite.hybrid.HybridHelper;

import java.nio.IntBuffer;

public class GLImage implements RawImage {
	private static final GLEngine engine = GLEngine.getInstance();
	private final int width, height;
	int tex;
	boolean glOriented = true;
	
	public GLImage( int width, int height) throws InvalidImageDimensionsExeption {
		if( width <= 0 || height <= 0)
			throw new InvalidImageDimensionsExeption("Invalid Image Dimensions");
		engine.glImageLoaded(this);
		GL2 gl = engine.getGL2();
		
		this.width = width;
		this.height = height;
		
		int result[] = new int[1];
		gl.glGenTextures(1, result, 0);
		this.tex = result[0];
        gl.glBindTexture( GLC.GL_TEXTURE_2D, tex);
        gl.glTexParameteri(GLC.GL_TEXTURE_2D,GLC.GL_TEXTURE_MIN_FILTER,GLC.GL_NEAREST);
        gl.glTexParameteri(GLC.GL_TEXTURE_2D,GLC.GL_TEXTURE_MAG_FILTER,GLC.GL_NEAREST);
        gl.glTexParameteri(GLC.GL_TEXTURE_2D,GLC.GL_TEXTURE_WRAP_S,GLC.GL_CLAMP_TO_EDGE);
        gl.glTexParameteri(GLC.GL_TEXTURE_2D,GLC.GL_TEXTURE_WRAP_T,GLC.GL_CLAMP_TO_EDGE);
        gl.glTexImage2D(GLC.GL_TEXTURE_2D,0,GL2.GL_RGBA8,
        		width, height, 0, GLC.GL_RGBA, GLC.GL_UNSIGNED_BYTE, null);
	}
	
	public GLImage( GLImage toCopy) {
		engine.glImageLoaded(this);
		GL2 gl = engine.getGL2();
		this.width = toCopy.width;
		this.height = toCopy.height;
		
		// Set the GL Target as the other image's texture and copy the data
		engine.setTarget(toCopy.tex);
		int result[] = new int[1];
		gl.glGenTextures(1, result, 0);
		this.tex = result[0];
        gl.glBindTexture( GLC.GL_TEXTURE_2D, tex);
        gl.glTexParameteri(GLC.GL_TEXTURE_2D,GLC.GL_TEXTURE_MIN_FILTER,GLC.GL_NEAREST);
        gl.glTexParameteri(GLC.GL_TEXTURE_2D,GLC.GL_TEXTURE_MAG_FILTER,GLC.GL_NEAREST);
        gl.glTexParameteri(GLC.GL_TEXTURE_2D,GLC.GL_TEXTURE_WRAP_S,GLC.GL_CLAMP_TO_EDGE);
        gl.glTexParameteri(GLC.GL_TEXTURE_2D,GLC.GL_TEXTURE_WRAP_T,GLC.GL_CLAMP_TO_EDGE);
        gl.glCopyTexImage2D(GLC.GL_TEXTURE_2D,0,GL2.GL_RGBA8,
        		0, 0, width, height, 0);
	}
	
	public GLImage( int texID, int width, int height, boolean glOriented) {
		engine.glImageLoaded(this);
		this.tex = texID;
		this.width = width;
		this.height = height;
		this.glOriented = glOriented;
	}
	
	@Override public int getWidth() { return width; }
	@Override public int getHeight() { return height; }
	public int getTexID() {return tex;}

	@Override
	public void flush() {
		if( tex != 0) {
			final int toDel = tex;
			tex = 0;
			
			// This is needed to make sure that it happens on the AWT Thread to prevent JOGL-internal deadlocks
			HybridHelper.queueToRun( () -> {
				engine.glImageUnloaded(this);
				GL2 gl = engine.getGL2();
				if( engine.getTarget() == toDel)
					engine.setTarget(0);
				gl.glDeleteTextures(1, new int[]{toDel}, 0);
			});
		}
	}
	
	@Override
	protected void finalize() throws Throwable {
		flush();
		super.finalize();
	}

	@Override
	public int getByteSize() {
		return width*height*4;
	}

	@Override
	public RawImage deepCopy() {
		return new GLImage( this);
	}

	@Override
	public GLGraphics getGraphics() {
		return new GLGraphics(this);
	}

	@Override
	public int getRGB(int x, int y) {
		if( x < 0 || y < 0 || x >= width || y >= height) return 0;
		GL2 gl = engine.getGL2();
		
		engine.setTarget(this);
		IntBuffer read = IntBuffer.allocate(1);
		gl.glReadnPixels( x, height-y-1, 1, 1, GL2.GL_BGRA, GL2.GL_UNSIGNED_INT_8_8_8_8_REV, 4, read);
		return read.get(0);
	}

	public void setGLOriented( boolean set) {this.glOriented = set;}
	@Override
	public boolean isGLOriented() {
		return glOriented;
	}

}

File: src/spirite/base/graphics/gl/GLImage.kt
package spirite.base.graphics.gl

import spirite.base.graphics.GraphicsContext
import spirite.base.graphics.RawImage
import spirite.base.graphics.RawImage.InvalidImageDimensionsExeption
import spirite.base.util.glu.GLC

class GLImage : RawImage {
    override val width : Int
    override val height: Int
    val engine: GLEngine
    override var isGLOriented: Boolean
    var tex : IGLTexture?
        get
        private set

    // region Constructors
    constructor( width: Int, height: Int, glEngine: GLEngine) {
        if( width <= 0 || height <= 0)
            throw InvalidImageDimensionsExeption("Invalid Image Dimensions")
        this.width = width
        this.height = height
        this.engine = glEngine
        this.isGLOriented = true

        val gl = glEngine.gl

        tex = gl.createTexture() ?: throw GLResourcException("Failed to create Texture")
        gl.bindTexture( GLC.TEXTURE_2D, tex)
        gl.texParameteri(GLC.TEXTURE_2D, GLC.TEXTURE_MIN_FILTER, GLC.NEAREST)
        gl.texParameteri(GLC.TEXTURE_2D, GLC.TEXTURE_MAG_FILTER, GLC.NEAREST)
        gl.texParameteri(GLC.TEXTURE_2D, GLC.TEXTURE_WRAP_S, GLC.CLAMP_TO_EDGE)
        gl.texParameteri(GLC.TEXTURE_2D, GLC.TEXTURE_WRAP_T, GLC.CLAMP_TO_EDGE)
        gl.texImage2D( GLC.TEXTURE_2D, 0, GLC.RGBA8, GLC.RGBA, GLC.UNSIGNED_BYTE,
                gl.createBlankTextureSource(width, height))
    }

    constructor( toCopy: GLImage) {
        width = toCopy.width
        height = toCopy.height
        engine = toCopy.engine
        isGLOriented = true

        val gl = engine.gl

        // Set the GL Target as the other image's texture and copy the data
        engine.target =  toCopy.tex
        tex = gl.createTexture() ?: throw GLResourcException("Failed to create Texture")
        gl.bindTexture(GLC.TEXTURE_2D, tex)
        gl.texParameteri(GLC.TEXTURE_2D, GLC.TEXTURE_MIN_FILTER, GLC.NEAREST)
        gl.texParameteri(GLC.TEXTURE_2D, GLC.TEXTURE_MAG_FILTER, GLC.NEAREST)
        gl.texParameteri(GLC.TEXTURE_2D, GLC.TEXTURE_WRAP_S, GLC.CLAMP_TO_EDGE)
        gl.texParameteri(GLC.TEXTURE_2D, GLC.TEXTURE_WRAP_T, GLC.CLAMP_TO_EDGE)
        gl.copyTexImage2D(GLC.TEXTURE_2D, 0, GLC.RGBA8, 0, 0, width, height, 0)
    }

    constructor( tex: IGLTexture, width: Int, height: Int, glEngine: GLEngine, isGLOriented: Boolean) {
        this.tex = tex
        this.width = width
        this.height = height
        this.engine = glEngine
        this.isGLOriented = isGLOriented
    }
    // endregion

    override val graphics: GraphicsContext get() = GLGraphics(this)
    override val byteSize: Int get() = width*height*4

    override fun flush() {
        val toDel = tex
        if( toDel != null) {
            // Must be run on the AWT Thread to prevent JOGL-internal deadlocks
            engine.runOnGLThread {
                //engine.glImageUnloaded(this)  // TODO
                if( engine.target == toDel)
                    engine.target = null
                engine.gl.deleteTexture(toDel)
            }
        }
    }

    override fun deepCopy(): RawImage = GLImage(this)

    override fun getRGB(x: Int, y: Int): Int {
        if (x < 0 || y < 0 || x >= width || y >= height) return 0
        val gl = engine.gl

        val read = gl.makeInt32Source(1)
        gl.readnPixels(x, if( isGLOriented) height-y-1 else y, 1, 1, GLC.BGRA, GLC.UNSIGNED_INT_8_8_8_8_REV, 4, read )
        return  read[0]
    }

    fun getRGBDirect( x: Int, y: Int): Int {
        if (x < 0 || y < 0 || x >= width || y >= height) return 0
        val gl = engine.gl

        val read = gl.makeInt32Source(1)
        gl.readnPixels(x, y, 1, 1, GLC.BGRA, GLC.UNSIGNED_INT_8_8_8_8_REV, 4, read )
        return  read[0]
    }

    protected fun finalize() {
        flush()
    }
}